File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0963_Minimum_Area_Rectangle_II.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Question_0963_Minimum_Area_Rectangle_II {
    public double minAreaFreeRect(int[][] points) {
        Map<String, List<int[]>> map = new HashMap<>();
        for (int i = 0; i < points.length; i++) {
            for (int j = i + 1; j < points.length; j++) {
                double x = (points[i][0] + points[j][0]) / 2.0;
                double y = (points[i][1] + points[j][1]) / 2.0;
                String key = x + "," + y + "," + getDist(points, i, j);
                map.putIfAbsent(key, new ArrayList<>());
                map.get(key).add(new int[]{i, j});
            }
        }
        double res = Double.MAX_VALUE;
        for (List<int[]> l : map.values()) {
            for (int i = 0; i < l.size(); i++) {
                for (int j = i + 1; j < l.size(); j++) {
                    res = Math.min(res, getArea(points, l.get(i), l.get(j)));
                }
            }
        }
        return res == Double.MAX_VALUE ? 0 : res;
    }

    private int getDist(int[][] p, int i, int j) {
        return (p[i][0] - p[j][0]) * (p[i][0] - p[j][0]) + (p[i][1] - p[j][1]) * (p[i][1] - p[j][1]);
    }

    private double getArea(int[][] p, int[] a, int[] b) {
        int dis1 = getDist(p, a[0], b[0]);
        int dis2 = getDist(p, a[0], b[1]);
        return Math.sqrt(dis1) * Math.sqrt(dis2);
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0963_Minimum_Area_Rectangle_II.cc
#include <vector>
#include <string>
#include <math.h>
#include <unordered_map>
#include <limits.h>

using namespace std;

class Solution {
 public:
  double minAreaFreeRect(vector<vector<int>>& points) {
    unordered_map<string, vector<vector<int>>> m;
    for (int i = 0; i < points.size(); i++) {
      for (int j = i + 1; j < points.size(); j++) {
        double x = (points[i][0] + points[j][0]) / 2.0;
        double y = (points[i][1] + points[j][1]) / 2.0;
        string key = to_string(x) + "," + to_string(y) + "," + to_string(getDist(points, i, j));
        m[key].push_back({i, j});
      }
    }
    double res = MAXFLOAT;
    for (auto& t : m) {
      for (int i = 0; i < t.second.size(); i++) {
        for (int j = i + 1; j < t.second.size(); j++) {
          res = min(res, getArea(points, t.second[i], t.second[j]));
        }
      }
    }
    return res == MAXFLOAT ? 0 : res;
  }

 private:
  int getDist(vector<vector<int>>& p, int i, int j) {
    return (p[i][0] - p[j][0]) * (p[i][0] - p[j][0]) + (p[i][1] - p[j][1]) * (p[i][1] - p[j][1]);
  }

  double getArea(vector<vector<int>>& p, vector<int>& a, vector<int>& b) {
    int dis1 = getDist(p, a[0], b[0]);
    int dis2 = getDist(p, a[0], b[1]);
    return sqrt(dis1) * sqrt(dis2);
  }
};
